vector/src/main/java/im/vector/app/features/settings/homeserver/HomeserverSettingsController.kt

```diff
@@ -0,0 +1,120 @@
+import im.vector.app.features.discovery.settingsSectionTitleItem
+import org.matrix.android.sdk.api.federation.FederationVersion
+import org.matrix.android.sdk.api.session.homeserver.HomeServerCapabilities
+import javax.inject.Inject
+
+class HomeserverSettingsController @Inject constructor(
+        private val stringProvider: StringProvider,
+        private val errorFormatter: ErrorFormatter
+) : TypedEpoxyController<HomeServerSettingsViewState>() {
+
+    var callback: Callback? = null
+
+    interface Callback {
+        fun retry()
+    }
+
+    override fun buildModels(data: HomeServerSettingsViewState?) {
+        data ?: return
+
+        buildHeader(data)
+        when (val federationVersion = data.federationVersion) {
+            is Loading,
+            is Uninitialized ->
+                loadingItem {
+                    id("loading")
+                }
+            is Fail          ->
+                errorWithRetryItem {
+                    id("error")
+                    text(errorFormatter.toHumanReadable(federationVersion.error))
+                    listener { callback?.retry() }
+                }
+            is Success       ->
+                buildFederationVersion(federationVersion())
+        }
+        buildCapabilities(data)
+    }
+
+    private fun buildHeader(state: HomeServerSettingsViewState) {
+        settingsCenteredImageItem {
+            id("icon")
+            drawableRes(R.drawable.ic_layers)
+        }
+        settingsSectionTitleItem {
+            id("urlTitle")
+            titleResId(R.string.hs_url)
+        }
+        settingsInfoItem {
+            id("urlValue")
+            helperText(state.baseUrl)
+        }
+    }
+
+    private fun buildFederationVersion(federationVersion: FederationVersion) {
+        settingsSectionTitleItem {
+            id("nameTitle")
+            titleResId(R.string.settings_server_name)
+        }
+        settingsInfoItem {
+            id("nameValue")
+            helperText(federationVersion.name)
+        }
+        settingsSectionTitleItem {
+            id("versionTitle")
+            titleResId(R.string.settings_server_version)
+        }
+        settingsInfoItem {
+            id("versionValue")
+            helperText(federationVersion.version)
+        }
+    }
+
+    private fun buildCapabilities(data: HomeServerSettingsViewState) {
+        settingsSectionTitleItem {
+            id("uploadTitle")
+            titleResId(R.string.settings_server_upload_size_title)
+        }
+
+        val limit = data.homeServerCapabilities.maxUploadFileSize
+
+        settingsInfoItem {
+            id("uploadValue")
+            if (limit == HomeServerCapabilities.MAX_UPLOAD_FILE_SIZE_UNKNOWN) {
+                helperTextResId(R.string.settings_server_upload_size_unknown)
+            } else {
+                helperText(stringProvider.getString(R.string.settings_server_upload_size_content, "${limit / 1048576L} MB"))
```

Comment:
It seems like a magic number at first glance. Can we use something like `1024.0.pow(2).toLong()` to avoid this? :)